File: ImasLiveDB/Domain/UseCases/OshiThemeResolution.swift
import Foundation

/// 担当(推し)カラーをアプリ全体テーマに使うときの、保存すべき値の解決。
///
/// 解決規則の本体は imas-core (Rust) の `domain/oshi_theme_resolution.rs` にある。
/// 「OFF のときは色だけ消して選択 ID は残す」「選択中の担当が外れていたら先頭へ
/// 黙って寄せる」という 2 つの非自明な判断とその理由もそちらに記載。
/// 結果型 `OshiThemeResolution` (idolId が nil = 現在の選択を変更しない) と
/// 射影型 `OshiThemePickIdol` は uniffi 生成バインディングが提供する。
///
/// ここは `Idol` を判定に要る 2 フィールド (id / color) の射影へ落として
/// 1 回の FFI 呼び出しへ委譲するだけの薄いラッパ。
/// 引数ラベルを `picks` にしているのは、生成バインディングの同名関数 (`pickIdols:`) と
/// ラベルまで同じにすると空配列リテラルで型推論が曖昧になるため。
func resolveOshiTheme(
    isEnabled: Bool,
    currentIdolId: String,
    picks: [Idol]
) -> OshiThemeResolution {
    resolveOshiTheme(
        isEnabled: isEnabled,
        currentIdolId: currentIdolId,
        pickIdols: picks.map { OshiThemePickIdol(id: $0.id, color: $0.color) }
    )
}

File: imas-core/src/inbound/oshi_theme_resolution.rs
//! 担当テーマ色解決の FFI 面。ロジックは domain::oshi_theme_resolution。
//!
//! アイドルのエンティティ全体は渡さず、射影 (`OshiThemePickIdol`) の列を受けて
//! 解決結果 (`OshiThemeResolution`) を 1 呼び出しで返す。担当は高々数十人なので
//! 射影列をそのまま渡しても境界コストは無視できる。

use crate::domain::oshi_theme_resolution::{OshiThemePickIdol, OshiThemeResolution};

/// 担当テーマ色を、現在の選択と担当一覧 (射影) から解決する。
/// `idol_id: None` は「現在の選択を変更しない」(OFF のとき)。
#[uniffi::export]
pub fn resolve_oshi_theme(
    is_enabled: bool,
    current_idol_id: String,
    pick_idols: Vec<OshiThemePickIdol>,
) -> OshiThemeResolution {
    crate::domain::oshi_theme_resolution::resolve_oshi_theme(
        is_enabled,
        &current_idol_id,
        &pick_idols,
    )
}
